### DISCUSSION TITLE
[FeatureRequest] Add the possibility to specify a time parameter in the speed component

### DISCUSSION THREAD

**Discussion by robin-gdwl** [visose/Robots#109] [Q&A] [answered]
Some Robot Languages (like UR) allow for specifying the time a movement should take. 
UR implements this with the `t=xx` parameter in most move fuctions. 
I saw that this functionality is already available in the `RobotCellUR` but I may be mistaken: 
https://github.com/visose/Robots/blob/9fc97b7ced60f321af310d68ee2cf1ea4b838ecd/Robots/RobotCellUR.cs#L342-L343  

would it be possible to expose a `time` input in the speed component or would that be confusing/break things for robots that don't use this functionality?

**Comment by visose** [comment, marked as answer]
You can create or modify a Speed parameter using a scripting component. You can there set variables not exposed as inputs such as Time and Acceleration

**Reply by v01tt** [reply]
Hi Visose, I tried using this method to access and set, the RotationalSpeed, on the UR3e robot. It would seem that changing this value has no affect on the "v=" value for the MovJ in the Program. I tried to use the Create Speed component., and am seeing the same result. The value just doesn't seem to change. It is always "v=0.628". See the code below:
![image](https://user-images.githubusercontent.com/77627544/154598030-e661b95d-677d-4416-885d-4f8d188f129f.png)
I want to change the "v=0.628", in line 8, which for the UR3e corresponds to the Joint Speed. AxisAccel, seems to be working and changes the "a=" values.  Any idea?

**Reply by visose** [reply]
RotationSpeed is the speed at which the TCP can rotate, not the speed of the joints. For all motions you can define the translation speed of the TCP and the rotation speed. I think rotation is not used by UR, only the translation speed is used. You don't need a scripting component to change this value.

In this plugin, the translation speed value (the main speed value), is used both for both joint and linear motions. This is the value that will affect the v=0.628. 

You might be wondering, how is this possible since in UR (and KUKA), the movej instruction expects a speed that corresponds to the speed of the leading axis in degrees per second, rather than a speed in distance per second. This angular speed value is approximated by the plugin by calculating how long it would take for the robot to reach that target if it was moving at the desired speed defined in distance per second, and then calculating the rotation speed of the leading axis that would result in the same time. The speed is always an approximation, as the TCP is not moving at constant speed in a straight line, but following an arc like motion.

In ABB robots, a joint motion instruction expects a distance speed rather than an angular speed, and this approximation is calculated directly by the controller.

**Reply by visose** [reply]
I have changed a few things about this, hopefully for the better. What I explained above is for all joint targets EXCEPT for the first one in the program, which happens to be the one you want to change. This is because the plugin doesn't know where your robot is before you run the program, so it cannot calculate how long it will take to reach your first target. This always defaulted to a safe speed of 10% of your lead axis speed.

In the next version (not published yet), you'll be able to control this initial speed by assuming that the maximum translation speed is 1000 mm/s. So if you set 250 mm/s, it will move at 1/4 of the lead axis speed. A speed of 100 mm/s (which is the default speed value) will behave as before (10% of the speed).

**Comment by robin-gdwl** [comment]
I will try that! Thank you for the quick response.

**Comment by robin-gdwl** [comment]
Could you give a quick overview how I can get access to the Robots-.NET library in the scripting component? I am having a hard time configuring the right namespaces etc.

Do you have an example script I could see as reference?

**Comment by robin-gdwl** [comment]
I made it work, with the help from this previous answer of yours: https://github.com/visose/Robots/issues/19#issuecomment-520556168 
Other examples would still be appreciated.

**Comment by robin-gdwl** [comment]
For Reference: this is the code I used to create a speed parameter with a time value:
Written in C#. I did not need to specify `using Robots;`. Adding the Assembly by right clicking on the c# scripting component was enough.

```cs
//using Robots;
//namespace Robots;

var inputspeed = new Speed() ;
inputspeed.Time = time;

Robotspeed = inputspeed;
Print(inputspeed.Time.ToString());
```
![image](https://user-images.githubusercontent.com/25942644/107245767-d08f6e00-6a2f-11eb-9cbd-80c0966ee1e5.png)

**Reply by lin-ycv** [reply]
I'm curious how accurate this is?
AFAIK the plugin does not simulate acceleration and deceleration, so if you define 30sec, the actual path will not be 30 sec?
can you share you experience so far? thanks

**Reply by visose** [reply]
The plugin does not do any path planning that is sent to the robot, it generates a high level robot program and the robot controller does the path planing. 

The robot program will contain a target that asks the controller to take 30 seconds to reach it. The real robot will of course accelerate and decelerate but will try to achieve those 30 seconds if its possible. As long as it doesn't have to keep correcting itself too much (tool weight and center of mass not set properly).

**Reply by robin-gdwl** [reply]
> I'm curious how accurate this is? 

As @visose said, the simulation in the plugin does not have an influence on the actual robot speed. 
The robot-code that is created tells the robot how fast the target should be reached. Then tries to achieve the motion in the specified time. 
In my experience this is quite accurate as long as the robot is physically able to move the distance in the specified time. Note that the robot will increase its speed and acceleration to meet the time constraint and might move incredibly quickly if the time is set to a low value. 
What kind of accuracy are you looking for? In my experience the motion is performed in the time that is specified, however the robot might do small pauses between motions that can differ +-0.2s.  If you need higher temporal accuracy a UR-robot might not be precise enough when controlled in this way.
